### PR TITLE
Fix for CUDA codegen

### PR DESCRIPTION
**Pull request by edopao**
This PR addresses #1388: fix python codegen and `SharedToGlobal1D` template to generate correct code for write without reduction.

### CODE REVIEW

File: dace/codegen/targets/cuda.py [hunk 1]
```diff
@@ -1128,26 +1128,42 @@ def _emit_copy(self, state_id, src_node, src_storage, dst_node, dst_storage, dst
                     accum = '::template Accum%s' % reduction_tmpl
 
                 if any(symbolic.issymbolic(s, sdfg.constants) for s in copy_shape):
-                    callsite_stream.write(('    {func}Dynamic<{type}, {bdims}, {is_async}>{accum}({args});').format(
+                    # reduction not yet supported by template for dynamic case
+                    if accum or custom_reduction:
+                        raise NotImplementedError(f'reduction not supported by template {funcname}')
```
Comment (tbennun) [inline dace/codegen/targets/cuda.py:1133]: Why is that the case? The C++ code allows it as far as I understand. Is it just a missing function?

File: dace/codegen/targets/cuda.py [hunk 2]
```diff
@@ -1128,26 +1128,42 @@ def _emit_copy(self, state_id, src_node, src_storage, dst_node, dst_storage, dst
...
                         bdims=', '.join(_topy(self._block_dims)),
                         is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'true',
-                        accum=accum,
-                        args=', '.join([src_expr] + _topy(src_strides) + [dst_expr] + custom_reduction +
+                        args=', '.join([src_expr] + _topy(src_strides) + [dst_expr] +
                                        _topy(dst_strides) + _topy(copy_shape))), sdfg, state_id, [src_node, dst_node])
+                elif funcname == 'dace::SharedToGlobal1D':
+                    # special case: use a new template struct that provides functions for copy and reduction
+                    callsite_stream.write(
+                        ('    {func}<{type}, {bdims}, {copysize}, {is_async}>{accum}({args});').format(
```
Comment (tbennun) [inline dace/codegen/targets/cuda.py:1144]: We're missing a test for each of the two cases.

File: dace/codegen/targets/cuda.py [hunk 3]
```diff
@@ -1128,26 +1128,42 @@ def _emit_copy(self, state_id, src_node, src_storage, dst_node, dst_storage, dst
...
+                             func=funcname,
+                             type=dst_node.desc(sdfg).dtype.ctype,
+                             bdims=', '.join(_topy(self._block_dims)),
+                             copysize=', '.join(_topy(copy_shape)),
+                             is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'true',
```
Comment (tbennun) [inline dace/codegen/targets/cuda.py:1149]: `'true' if ... else 'true'`: was this the intent?

File: dace/runtime/include/dace/math.h [hunk 4]
```diff
@@ -526,7 +526,7 @@ namespace dace
         template<typename T>
         DACE_CONSTEXPR DACE_HDFI int ifloor(const T& a)
         {
-            return (int)std::floor(a);
+            return (int)std::floor(static_cast<double>(a));
```
Comment (tbennun) [inline dace/runtime/include/dace/math.h:529]: that will slow down float code. I cannot approve that. Perhaps specialization for specific types is better here.

**Review by tbennun — CHANGES_REQUESTED** [review]

**Review by edopao — COMMENTED** [review]
Thank you the review, I will address your comments in a new commit.

**Comment by edopao** [comment]
@tbennun Test added, please re-review.

File: dace/codegen/targets/cuda.py [hunk 5]
```diff
@@ -1145,7 +1157,7 @@ def _emit_copy(self, state_id, src_node, src_storage, dst_node, dst_storage, dst
                              bdims=', '.join(_topy(self._block_dims)),
                              copysize=', '.join(_topy(copy_shape)),
                              dststrides=', '.join(_topy(dst_strides)),
-                             is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'true',
+                             is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'false',
```
Comment (tbennun) [inline dace/codegen/targets/cuda.py:1160]: Same

**Review by tbennun — CHANGES_REQUESTED** [review]
Thank you. Only minor comments remain

File: dace/codegen/targets/cuda.py [hunk 6]
```diff
@@ -1132,10 +1132,22 @@ def _emit_copy(self, state_id, src_node, src_storage, dst_node, dst_storage, dst
                         func=funcname,
                         type=dst_node.desc(sdfg).dtype.ctype,
                         bdims=', '.join(_topy(self._block_dims)),
-                        is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'true',
+                        is_async='true' if state_dfg.out_degree(dst_node) > 0 else 'false',
```
[the remark this answers is not in the document]
Comment (edopao) [reply dace/codegen/targets/cuda.py]: Correct. I did not pay enough attention to `is_async` before. Besides correcting the value of this argument, I have also moved the synchronization point in the template function after the thread-level copy (see my last commit on copy.cuh).

**Comment by edopao** [comment]
@tbennun Thank you for the review. As I commented above, I have done one additional change related to `is_async` flag.

**Comment by edopao** [comment]
Gentle reminder for @tbennun: please check my last comment and whether you can approve this PR.

**Review by tbennun — APPROVED** [review]

**Pull request merged**